src/Primitives.hs

```diff
@@ -221,53 +217,63 @@ primitiveImplements x@(XObj _ _ _) ctx args =
-          pure (ctx {contextTypeEnv = TypeEnv (envInsertAt (getTypeEnv typeEnv) (getPath annXObj) (Binder adjustedMeta annXObj))})
-        XObj (Lst (XObj (Deftype _) _ _ : _)) _ _ ->
-          pure (ctx {contextTypeEnv = TypeEnv (envInsertAt (getTypeEnv typeEnv) (getPath annXObj) (Binder adjustedMeta annXObj))})
-        XObj (Lst (XObj (DefSumtype _) _ _ : _)) _ _ ->
-          pure (ctx {contextTypeEnv = TypeEnv (envInsertAt (getTypeEnv typeEnv) (getPath annXObj) (Binder adjustedMeta annXObj))})
-        _ ->
-          do
-            when (projectEchoC proj) $
-              putStrLn (toC All (Binder emptyMeta annXObj))
-            case previousType of
-              Just previousTypeUnwrapped ->
-                unless (areUnifiable (forceTy annXObj) previousTypeUnwrapped) $
-                  do
-                    emitWarning
-                      ( "Definition at " ++ prettyInfoFromXObj annXObj ++ " changed type of '" ++ show (getPath annXObj)
-                          ++ "' from "
-                          ++ show previousTypeUnwrapped
-                          ++ " to "
-                          ++ show (forceTy annXObj)
-                      )
-              Nothing -> pure ()
-            case Meta.get "implements" previousMeta of
-              Just (XObj (Lst interfaces) _ _) ->
-                do
-                  let result = foldM (\ctx' (xobj, interface) -> registerInInterface ctx' xobj interface) ctx (zip (cycle [annXObj]) (map getPath interfaces))
-                  case result of
-                    Left err ->
-                      do
-                        case contextExecMode ctx of
-                          Check ->
-                            let fppl = projectFilePathPrintLength (contextProj ctx)
-                             in putStrLn (machineReadableInfoFromXObj fppl annXObj ++ " " ++ err)
-                          _ -> putStrLnWithColor Red err
-                        pure ctx
-                    Right ctx' -> pure (ctx' {contextGlobalEnv = envInsertAt globalEnv (getPath annXObj) (Binder adjustedMeta annXObj)})
-              _ -> pure (ctx {contextGlobalEnv = envInsertAt globalEnv (getPath annXObj) (Binder adjustedMeta annXObj)})
+  pure (hideIt freshBinder)
+    >>= \newBinder -> case annXObj of
+      XObj (Lst (XObj (Defalias _) _ _ : _)) _ _ ->
+        pure (ctx {contextTypeEnv = TypeEnv (envInsertAt (getTypeEnv typeEnv) (getPath annXObj) newBinder)})
+      XObj (Lst (XObj (Deftype _) _ _ : _)) _ _ ->
+        pure (ctx {contextTypeEnv = TypeEnv (envInsertAt (getTypeEnv typeEnv) (getPath annXObj) newBinder)})
+      XObj (Lst (XObj (DefSumtype _) _ _ : _)) _ _ ->
+        pure (ctx {contextTypeEnv = TypeEnv (envInsertAt (getTypeEnv typeEnv) (getPath annXObj) newBinder)})
+      _ -> defineGlobal
+  where
+    freshBinder = (Binder emptyMeta annXObj)
+    defineGlobal :: IO Context
+    defineGlobal =
+      when (projectEchoC proj) (putStrLn (toC All (Binder emptyMeta annXObj)))
```

Comment:
I feel like this function is very imperative in it's style (with printing and all), so that `do` notation actually could make more sense. I'm open to leaving it as-is, but please consider it :)

Reply:
Sure thing!